protocol/consistencychecks.go

```diff
@@ -42,222 +38,243 @@ type ConsistencyChecks struct {
 // the pinning directory's STR at epoch 0 or
 // the consistency state read from a persistent storage.
 func NewCC(savedSTR *m.SignedTreeRoot, useTBs bool, signKey sign.PublicKey) *ConsistencyChecks {
-
 	// TODO: see #110
 	if !useTBs {
-		panic("Currently the server is forced to use TBs")
+		panic("[coniks] Currently the server is forced to use TBs.")
+	}
+	if savedSTR == nil {
+		panic("[coniks] Expect a non-nil consistency state.")
 	}
-
 	cc := &ConsistencyChecks{
+		oldSTR:   savedSTR,
 		SavedSTR: savedSTR,
 		useTBs:   useTBs,
 		signKey:  signKey,
 	}
-
 	if useTBs {
-		cc.TBs = make(map[string]*m.TemporaryBinding)
+		cc.TBs = make(map[string]*TemporaryBinding)
 	}
 	return cc
 }
 
-// UpdateConsistency verifies the directory's response for a request and
-// updates the current state if all the verifications are passed.
+// HandleResponse verifies the directory's response for a request.
 // It first verifies the directory's returned status code of the request.
 // If the status code is not in the ErrorResponses array, it means
 // the directory has successfully handled the request.
 // The verifier will then verify the consistency state of the response.
-func (cc *ConsistencyChecks) UpdateConsistency(requestType int, msg *Response,
+// This will panic if it is called with an int
+// which isn't a valid/known request-type.
+// Note that the consistency state would be updated regardless of
+// whether the checks pass / fail, since it contains proof of being issued.
+func (cc *ConsistencyChecks) HandleResponse(requestType int, msg *Response,
 	uname string, key []byte) ErrorCode {
 	if ErrorResponses[msg.Error] {
 		return msg.Error
 	}
+	switch requestType {
+	case RegistrationType, KeyLookupType:
+		if _, ok := msg.DirectoryResponse.(*DirectoryProof); !ok {
+			return ErrorMalformedDirectoryMessage
+		}
+	default:
+		panic("[coniks] Unknown request type.")
+	}
+	if err := cc.updateSTR(requestType, msg); err != nil {
+		return err.(ErrorCode)
+	}
+	if err := cc.checkConsistency(requestType, msg, uname, key); err != Passed {
+		return err
+	}
+	if err := cc.updateTBs(requestType, msg, uname, key); err != nil {
+		return err.(ErrorCode)
+	}
+	return Passed
+}
 
+func (cc *ConsistencyChecks) updateSTR(requestType int, msg *Response) error {
 	var str *m.SignedTreeRoot
-	var err error
+	switch requestType {
+	case RegistrationType, KeyLookupType:
+		str = msg.DirectoryResponse.(*DirectoryProof).STR
+		// FIXME: check whether the STR was issued on time and whatnot.
+		// Maybe it has something to do w/ #81 and client transitioning between epochs.
+		// Try to verify w/ what's been saved
+		if err := cc.verifySTR(str); err == nil {
```

Comment:
80e453caa501ea639c149e7af8f507b3ed6b0ed4 makes me feel like a lazy developer can always pass a nil pointer of `savedSTR` instead. 

Since we don't have #119 yet, so yes, I agree it should be reverted too.